File: sliding-window/repeated-dna-sequence.py
def find_repeated_sequences(s, k):
    window_size = k
    seq_len = len(s)

    output = set()
    if seq_len < window_size:
        return output

    seen = set()
    base = 4

    mapping = {"A": 1, "G": 2, "C": 3, "T": 4}
    nums = []
    for i in range(seq_len):
        nums.append(mapping.get(s[i]))

    hash_value = 0

    for start in range(seq_len - window_size + 1):
        if start == 0:
            for end in range(window_size):
                hash_value = hash_value * base + nums[end]
        else:
            hash_value = (
                hash_value - (nums[start - 1] * pow(base, window_size - 1))
            ) * base + nums[start + window_size - 1]

        if hash_value in seen:
            output.add(s[start : start + window_size])
        else:
            seen.add(hash_value)

    return output

File: sliding-window/repeated-dna-sequence.js
export function findRepeatedSequences(s, k) {
	let n = s.length;

	let outputSet = new Set();
	if (n < k) return outputSet;

	let seenSet = new Set();

	let nMap = new Map([['A',1],['G',2],['T',3],['C',4]]);

	let nums = new Array(n);
	for (let i = 0; i < n; i++) {
			nums[i] = nMap.get(s[i]);
	}

	let a = 4;
	let hashValue = 0;

	for (let i = 0; i < n - k + 1; i++) {
			if (i === 0) {
					for (let j = 0; j < k; j++) {
							hashValue = hashValue * a + nums[j];
					}
			} else {
					hashValue = (hashValue - (nums[i - 1] * Math.pow(a, k - 1))) * a + nums[i + k - 1];
			}

			if (seenSet.has(hashValue)) {
					outputSet.add(s.substring(i, i + k));
			} else {
					seenSet.add(hashValue);
			}
	}

	return outputSet;
}
